### PR TITLE
[backport #1727] SI-7359 cyclic nested java class

### PR DESCRIPTION
**Pull request by adriaanm**
The original commit message:

SI-6548 reflection correctly enters jinners

When completing Java classes, runtime reflection enumerates their
fields, methods, constructors and inner classes, loads them and
enters them into either the instance part (ClassSymbol) or the
static part (ModuleSymbol).

However unlike fields, methods and constructors, inner classes don't
need to be entered explicitly - they are entered implicitly when
being loaded.

This patch fixes the double-enter problem, make sure that enter-on-load
uses the correct owner, and also hardens jclassAsScala against double
enters that can occur in a different scenario.

Since the fix is about Java-compiled classes, the test needs *.class
artifacts produced by javac. Therefore I updated javac-artifacts.jar to
include the new artifacts along with their source code.

review by @xeno-by

### CODE REVIEW

**Comment by xeno-by** [comment]
I would probably change the description, since it refers to `javac-artifacts.jar` which have been removed since then.

**Comment by xeno-by** [comment]
It would also be helpful to link to the commit being backported.

**Comment by xeno-by** [comment]
Other than that LGTM

**Comment by adriaanm** [comment]
updated, please also note the changed check file

File: test/files/run/t6989/JavaClass_1.java [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
 // I'm leaving the incorrect results of FromJavaClassCompleters in the check
 // file, so that we get notified when something changes there.
+// ^^^ It's not clear what those incorrect results were, but the fix for SI-7359
+// (backport of fix for SI-6548) has probably resolved some of these. OP, please revisit this comment.
```
Comment (xeno-by) [inline test/files/run/t6989/JavaClass_1.java:11]: Incorrectness of the results is explained here: https://github.com/adriaanm/scala/commit/b43ae58f312e4769de5a6942b4cb820cd609f31b

**Comment by xeno-by** [comment]
Okay the change in the check file is explained by this commit: https://github.com/adriaanm/scala/commit/85b63b81d5951a78547641e3feab0886f6013ea1. Back then this fix was already merged into 2.10.x, but you reverted it because of compatibility constraints, which made inner classes to be entered twice, polluting the check file. Now inner classes are entered once, which left a dent in the check file, bringing it to the usual slim size.

File: test/files/run/t6989/JavaClass_1.java (hunk 1 above)
Comment (adriaanm) [inline test/files/run/t6989/JavaClass_1.java:11]: I saw that comment, but what does it mean? Is it referring to the duplicates? The owners? The constructors? What's expected? What's fixed after this commit.

Comment (xeno-by) [inline test/files/run/t6989/JavaClass_1.java:11]: Duplicates were introduced by a later revert, so they weren't in the checkfile when the message was written. I guess incorrectness I alluded to comes from SI-7071 and SI-7072.

**Comment by adriaanm** [comment]
bah, forgot check file

**Pull request merged**